File: src/main/java/club/libridge/libridgebackend/core/Auction.java
package club.libridge.libridgebackend.core;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import club.libridge.libridgebackend.core.exceptions.AuctionAlreadyFinishedException;
import club.libridge.libridgebackend.core.exceptions.CallInAnotherPlayersTurnException;
import club.libridge.libridgebackend.core.exceptions.InsufficientBidException;
import club.libridge.libridgebackend.core.exceptions.InvalidCallException;
import lombok.Getter;
import lombok.NonNull;

/**
 * The rules for Bridge Auction are as follows (paraphrasing/quoting the 2017 Laws of Bridge - LAWS 17-22):
 *
 * - The first call is made by the dealer - Every other call is made by the player who goes next from the last caller - The auction ends after 4
 * consecutive passes (this only happens when they are the first 4 calls) - The auction ends after 3 consecutive passes if a bid has already been
 * made.
 *
 * - Pass is always a valid call - Double (X) can only be called if and only if the last non-pass call was made by the opponents AND it was a bid. -
 * Redouble (XX) can only be called if and only if the last non-pass call was made by the opponents AND it was a double (X). - The first bid is freely
 * chosen from all bids. All other bids must supersede the last bid (and consequently all previous ones).
 *
 */

public final class Auction {

    private static final int NUMBER_OF_INITIAL_PASSES_TO_END_AUCTION = 4;
    private static final int NUMBER_OF_PASSES_TO_END_AUCTION_AFTER_BID = 3;

    @Getter
    private final Direction dealer;
    @Getter
    private Direction currentTurn;
    private final List<Call> bids;
    @Getter
    private boolean finished;
    /**
     * These are initialized as -1
     */
    private int lastNonPassCallIndex;
    private int lastBidIndex;
    private Contract finalContract;

    public Auction(@NonNull Direction dealer) {
        this.dealer = dealer;
        this.currentTurn = dealer;
        this.bids = new ArrayList<Call>();
        this.finished = false;
        this.lastNonPassCallIndex = -1;
        this.lastBidIndex = -1;
        this.finalContract = null;
    }

    public List<Call> getBids() {
        return Collections.unmodifiableList(bids);
    }

    public void makeCall(@NonNull Direction direction, @NonNull Call call) {
        if (this.isFinished()) {
            throw new AuctionAlreadyFinishedException();
        }
        if (this.currentTurn != direction) {
            throw new CallInAnotherPlayersTurnException();
        }
        if (direction == null || call == null) {
            throw new IllegalArgumentException("Parameters should not be null");
        }

        if (call.isPass()) {
            addCall(call);
        } else if (call.isDouble()) {
            if (isDoubleValid()) {
                addCall(call);
            } else {
                throw new InvalidCallException();
            }
        } else if (call.isRedouble()) {
            if (isRedoubleValid()) {
                addCall(call);
            } else {
                throw new InvalidCallException();
            }
        } else if (call.isBid()) {
            if (lastBidIndex == -1) {
                addCall(call);
            } else {
                Bid lastBid = (Bid) this.bids.get(lastBidIndex);
                Bid currentBid = (Bid) call;
                if (currentBid.compareTo(lastBid) > 0) {
                    addCall(call);
                } else {
                    throw new InsufficientBidException();
                }
            }
        }

        checkAndUpdateForFinishedAuction();
    }

    private boolean isDoubleValid() {
        return lastNonPassCallIndex >= 0 && isLastNonPassBidFromOpponents() && this.bids.get(lastNonPassCallIndex).isBid();
    }

    private boolean isRedoubleValid() {
        return lastNonPassCallIndex >= 0 && isLastNonPassBidFromOpponents() && this.bids.get(lastNonPassCallIndex).isDouble();
    }

    private boolean isLastNonPassBidFromOpponents() {
        if (lastNonPassCallIndex < 0) {
            return false;
        }
        return (lastNonPassCallIndex - this.bids.size()) % 2 != 0;
    }

    private void addCall(@NonNull Call call) {
        Call unmodifiableCopy = getCallFromBiddingBox(call);
        this.bids.add(unmodifiableCopy);
        this.currentTurn = this.currentTurn.next();
        if (!unmodifiableCopy.isPass()) {
            this.lastNonPassCallIndex = this.bids.size() - 1;
        }
        if (unmodifiableCopy.isBid()) {
            this.lastBidIndex = this.bids.size() - 1;
        }
    }

    private Call getCallFromBiddingBox(@NonNull Call call) {
        return BiddingBox.get(call);
    }

    private void checkAndUpdateForFinishedAuction() {
        int numberOfTailPasses = this.getTailPasses();
        if (numberOfTailPasses >= NUMBER_OF_INITIAL_PASSES_TO_END_AUCTION) {
            this.finished = true;
        } else if (numberOfTailPasses == NUMBER_OF_PASSES_TO_END_AUCTION_AFTER_BID && lastBidIndex >= 0) {
            this.finished = true;
        }
    }

    private int getTailPasses() {
        int tailPasses = 0;
        for (int i = this.bids.size() - 1; i >= 0; i--) {
            if (this.bids.get(i).isPass()) {
                tailPasses++;
            } else {
                break;
            }
        }
        return tailPasses;
    }

    public Contract getFinalContract() {
        if (!this.isFinished()) {
            throw new IllegalStateException("Auction must be finished to have a final contract.");
        }

        if (this.finalContract != null) {
            return this.finalContract;
        }

        Bid lastBid = null;
        int lastBidIndex = -1;
        for (int i = this.bids.size() - 1; i >= 0; i--) {
            if (this.bids.get(i).isBid()) {
                lastBid = (Bid) this.bids.get(i);
                lastBidIndex = i;
                break;
            }
        }

        if (lastBid == null) {
            // Ignoring PASS for now
            throw new IllegalStateException("PASS is not implemented yet.");
        } else {
            List<Call> callsAfterLastBid = this.bids.subList(lastBidIndex, this.bids.size());
            boolean redoubled = callsAfterLastBid.stream().anyMatch(call -> call.isRedouble());
            boolean doubled = callsAfterLastBid.stream().anyMatch(call -> call.isDouble());
            PenaltyStatus penaltyStatus = PenaltyStatus.NONE;
            if (doubled) {
                penaltyStatus = PenaltyStatus.DOUBLED;
            } else if (redoubled) {
                penaltyStatus = PenaltyStatus.REDOUBLED;
            }
            boolean nonVul = false;
            OddTricks oddTricks = lastBid.getOddTricks();
            Strain strain = lastBid.getStrain();
            this.finalContract = new Contract(oddTricks, strain, penaltyStatus, nonVul);
            return this.finalContract;
        }
    }

    @Override
    public String toString() {
        StringBuilder returnValue = new StringBuilder();
        List<Call> calls = this.getBids();
        for (Call call : calls) {
            if (call.isPass()) {
                returnValue.append("P ");
            } else if (call.isDouble()) {
                returnValue.append("X ");
            } else if (call.isRedouble()) {
                returnValue.append("XX ");
            } else {
                Bid bid = (Bid) call;
                returnValue.append(bid.getOddTricks().getSymbol() + bid.getStrain().getSymbol() + " ");
            }
        }
        return returnValue.toString();
    }

}

File: src/main/scala/scalabridge/Auction.scala
package scalabridge

import scalabridge.exceptions.auction.*

import scala.jdk.javaapi.CollectionConverters
import scala.util.Failure
import scala.util.Success
import scala.util.Try

/**
  * The rules for Bridge Auction are as follows (paraphrasing/quoting the 2017 Laws of Bridge - LAWS 17-22):
  * 
  * - The first call is made by the dealer
  * - Every other call is made by the player who goes next from the last caller
  * - The auction ends after 4 consecutive passes (this only happens when they are the first 4 calls)
  * - The auction ends after 3 consecutive passes if a bid has already been made.
  * 
  * - Pass is always a valid call
  * - Double (X) can be called if and only if the last non-pass call was made by the opponents AND it was a bid.
  * - Redouble (XX) can be called if and only if the last non-pass call was made by the opponents AND it was a double (X).
  * - The first bid is freely chosen from all bids. All other bids must supersede the last bid (and consequently all previous ones).
  *
  */
case class Auction(
    dealer: Direction,
    calls: List[Call] = List.empty
) extends Validated[Auction] {

  override def getValid(): Either[Iterable[Throwable], Auction] = {
    val intermediateCalls = calls.zipWithIndex
      .map((callToTry, index) =>
        (
          Auction(this.dealer, calls.drop(index + 1)),
          callToTry,
          this.currentTurn.next((index + 1) * 3)
        )
      )
    val failures = intermediateCalls
      .map((auction, callToTry, directionToTry) => auction.makeCall(directionToTry, callToTry))
      .collect { case Failure(throwable) => throwable }
    if (failures.isEmpty) Right(this)
    else Left(failures)
  }

  import Auction.*

  def this(dealer: Direction) = this(dealer, List.empty)

  private lazy val lastNonPassCall: Option[(Call, Direction)] = {
    val callWithIndexOption = calls.zipWithIndex.find(!_._1.isPass)
    callWithIndexOption match
      case None => None
      case Some((call, index)) => {
        val positionsFromDealer = calls.size - index - 1
        Some(call, dealer.next(positionsFromDealer))
      }
  }
  private lazy val lastBidOption: Option[Bid] = calls.collectFirst({ case x: Bid =>
    x.asInstanceOf[Bid]
  })

  private val currentTurn: Direction = dealer.next(calls.size)

  private def isDoubleValid = lastNonPassCall match
    case None                    => false
    case Some((call, direction)) => call.isBid && directionIsOpponent(direction)

  private def isRedoubleValid = lastNonPassCall match
    case None                    => false
    case Some((call, direction)) => call.isDouble && directionIsOpponent(direction)

  private def directionIsOpponent(direction: Direction) =
    Side.getFromDirection(direction) != Side.getFromDirection(currentTurn)

  private def copyWithCall(call: Call) = this.copy(calls = call +: this.calls)

  val isFinished: Boolean = calls.equals(numberOfInitialPassesToEndAuction) ||
    (calls.startsWith(numberOfConsecutivePassesToEndAuction) && lastNonPassCall.isDefined)

  def getCalls(): java.util.List[Call] = CollectionConverters.asJava(calls)
  def makeCall(direction: Direction, call: Call): Try[Auction] = {
    if (this.isFinished) Failure(new AuctionAlreadyFinishedException)
    else if (currentTurn != direction) Failure(new CallInAnotherPlayersTurnException)
    else
      call match
        case PassingCall => Success(copyWithCall(call))
        case DoubleCall =>
          if (isDoubleValid) Success(copyWithCall(call)) else Failure(new InvalidCallException)
        case RedoubleCall =>
          if (isRedoubleValid) Success(copyWithCall(call)) else Failure(new InvalidCallException)
        case bid: Bid =>
          lastBidOption match
            case None => Success(copyWithCall(call))
            case Some(lastBid) =>
              if (bid.supersedes(lastBid)) Success(copyWithCall(call))
              else Failure(new InsufficientBidException)
  }
  def makeCallSafe(direction: Direction, call: Call): Auction =
    makeCall(direction, call).getOrElse(this)
  def getFinalContract: Option[Contract] = {
    if (!isFinished) None
    else {
      lastBidOption match
        case None => Some(AllPassContract)
        case Some(lastBid) => {
          val callsAfterLastBid = calls.span(_ != lastBid)._1
          val isDoubled = callsAfterLastBid.contains(DoubleCall)
          val isRedoubled = callsAfterLastBid.contains(RedoubleCall)
          val penaltyStatus =
            if (isRedoubled) PenaltyStatus.REDOUBLED
            else if (isDoubled) PenaltyStatus.DOUBLED
            else PenaltyStatus.NONE
          Some(Contract(lastBid.oddTricks, lastBid.strain, penaltyStatus))
        }
    }
  }
}
object Auction {
  def apply(dealer: Direction) = new Auction(dealer)
  private val numberOfInitialPassesToEndAuction =
    List(PassingCall, PassingCall, PassingCall, PassingCall)
  private val numberOfConsecutivePassesToEndAuction = List(PassingCall, PassingCall, PassingCall)
}
